codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/RestJsonProtocolGenerator.java

```diff
@@ -215,6 +215,12 @@ protected void writeMiddlewareDocumentDeserializerDelegator(
         writer.addUseImports(SmithyGoDependency.JSON);
         writer.write("decoder := json.NewDecoder(body)");
         writer.write("decoder.UseNumber()");
+        writer.write("var shape interface{}");
+        writer.addUseImports(SmithyGoDependency.IO);
+        writer.openBlock("if err := decoder.Decode(&shape); err != nil && err != io.EOF {", "}", () -> {
+            writer.addUseImports(SmithyGoDependency.SMITHY);
+            writer.write("return out, metadata, &smithy.DeserializationError{Err: err}");
+        });
```

Comment:
not significant, but can that be shared across the implementations, it looks like this duplicated several times.

Sometimes it's deserialization error has a fmt.Errorf and sometimes it doesn't, probably want consistent there where it makes sense.

Reply:
There are some subtitles here due to how de-serialization expectations are different about returns for errors versus returns from the document body path. I will refactor where possible.